### ISSUE TITLE
Suggestion: Allow skipping the properties of type `void`

### ISSUE THREAD

**Issue by emirotin** [Microsoft/TypeScript#15189]
**TypeScript Version:**  2.3.0

**Code**

I'd like this to be a valid TS:

```ts
interface T {
  a: number;
  b: void;
}

const x: T = { a: 1 }
```

**Why would I?**

This contrived example doesn't make much sense. Why would someone define a property of type `void` and then be willing to skip it, right?

Here's a lengthy explanation of my problem: http://stackoverflow.com/questions/43387510/typescript-for-generic-type-interface-make-the-property-required-only-if-its

Long story short I have a generic interface, whenever the type parameter is a _real type_ I want the property to be required, so I cannot mark it as optional.
But if the type is `void` (and I do use 2.3.0's default type parameters feature to set it and thus make the type parameter itself optional) I don't want to keep typing `b: undefined` everywhere.

**Impact on emitted code**: none, it only touches the TS compile-time validation.

**Comment by mhegazy** [comment]
it seems to me that your type `T` should be defined as:

```ts
type T<P> = {a: number } | {a: number, b: P};
```

**Comment by emirotin** [comment]
That wouldn't do the trick because it [doesn't](https://www.typescriptlang.org/play/index.html#src=type%20T%3CP%3E%20%3D%20%7Ba%3A%20number%20%7D%20%7C%20%7Ba%3A%20number%2C%20b%3A%20P%7D%3B%0A%0Avar%20x%3A%20T%3Cnumber%3E%20%20%3D%20%7B%20a%3A%201%20%7D%20%2F%2F%20it's%20valid%2C%20but%20I%20want%20it%20to%20fail) enforce `b` when `P` isn't `void`:

**Comment by mhegazy** [comment]
then yous should have two types:

```ts
type PlainAction = { a: number };
type ActionWithPayload<T> = { a: number, b: T };
type AnyAction<T> = PlainAction | ActionWithPayload<T>;

const x: ActionWithPayload<number> = { a: 1 }; // Error
```

**Comment by emirotin** [comment]
You're not making use of the `AnyAction` type which is as useless as the previous example — [doesn't](https://www.typescriptlang.org/play/index.html#src=type%20PlainAction%20%3D%20%7B%20a%3A%20number%20%7D%3B%0Atype%20ActionWithPayload%3CT%3E%20%3D%20%7B%20a%3A%20number%2C%20b%3A%20T%20%7D%3B%0Atype%20AnyAction%3CT%3E%20%3D%20PlainAction%20%7C%20ActionWithPayload%3CT%3E%3B%0A%0Aconst%20x%3A%20AnyAction%3Cnumber%3E%20%3D%20%7B%20a%3A%201%20%7D#src=type%20PlainAction%20%3D%20%7B%20a%3A%20number%20%7D%3B%0Atype%20ActionWithPayload%3CT%3E%20%3D%20%7B%20a%3A%20number%2C%20b%3A%20T%20%7D%3B%0Atype%20AnyAction%3CT%3E%20%3D%20PlainAction%20%7C%20ActionWithPayload%3CT%3E%3B%0A%0Aconst%20x%3A%20AnyAction%3Cnumber%3E%20%3D%20%7B%20a%3A%201%20%7D%3B%20%2F%2F%20Error%0A%0A) enforce `b`.

The linked SO question explains why having two separate types is impractical.

**Comment by gcanti** [comment]
@emirotin If it can be helpful there's a possible workaround using a bit of type level programming

```ts
// type level boolean
type Bool = 'true' | 'false'
// type level if
type If<Test extends Bool, Then, Else> = {
  true: Then,
  false: Else
}[Test]

// definition

type Void = { isVoid: 'true', type: void }
type Type<T> = { isVoid: 'false', type: T }
type AnyType = Void | Type<any>

// usage

type BaseAction = { type: string }

type Action<T extends AnyType> = If<
  T['isVoid'],
  BaseAction,
  BaseAction & { payload: T['type'] }
  >

// tests

const action1: Action<Void> = { type: 'action1' }
const action2: Action<Type<number>> = { type: 'action1', payload: 2 }
const action3: Action<Type<number>> = { type: 'action1', payload: '2' } // error
const action4: Action<Type<number>> = { type: 'action1' } // error
const action5: Action<Void> = { type: 'action1', payload: undefined } // error
```

**Comment by emirotin** [comment]
Wow @gcanti this is rad :) The `If` part with immediate object access is something I would never expect to work.

Obviously I'd like my code to work with "native" types without the custom Type / Void wrappers, but anyway this is an awesome lesson, thank you.

**Comment by emirotin** [comment]
Hm, actually this doesn't seem to work: the `Test` seems to be generalized to `Then | Else`, and also `any` seems to be plaguing the entire codebase preventing it from being specific (but the latter is easy to fix): [playground](https://www.typescriptlang.org/play/index.html#src=%2F%2F%20type%20level%20boolean%0Atype%20Bool%20%3D%20'true'%20%7C%20'false'%0A%2F%2F%20type%20level%20if%0Atype%20If%3CTest%20extends%20Bool%2C%20Then%2C%20Else%3E%20%3D%20%7B%0A%20%20true%3A%20Then%2C%0A%20%20false%3A%20Else%0A%7D%5BTest%5D%0A%0A%2F%2F%20definition%0A%0Atype%20Void%20%3D%20%7B%20isVoid%3A%20'true'%2C%20type%3A%20void%20%7D%0Atype%20Type%3CT%3E%20%3D%20%7B%20isVoid%3A%20'false'%2C%20type%3A%20T%20%7D%0Atype%20AnyType%20%3D%20Void%20%7C%20Type%3Cany%3E%0A%0A%2F%2F%20usage%0A%0Atype%20BaseAction%20%3D%20%7B%20type%3A%20string%20%7D%0A%0Atype%20Action%3CT%20extends%20AnyType%3E%20%3D%20If%3C%0A%20%20T%5B'isVoid'%5D%2C%0A%20%20BaseAction%2C%0A%20%20BaseAction%20%26%20%7B%20payload%3A%20T%5B'type'%5D%20%7D%0A%20%20%3E%0A%0A%2F%2F%20tests%0A%0Aconst%20action1%3A%20Action%3CVoid%3E%20%3D%20%7B%20type%3A%20'action1'%20%7D%0Aconst%20action2%3A%20Action%3CType%3Cnumber%3E%3E%20%3D%20%7B%20type%3A%20'action1'%2C%20payload%3A%202%20%7D%0Aconst%20action3%3A%20Action%3CType%3Cnumber%3E%3E%20%3D%20%7B%20type%3A%20'action1'%2C%20payload%3A%20'2'%20%7D%20%2F%2F%20error%0Aconst%20action4%3A%20Action%3CType%3Cnumber%3E%3E%20%3D%20%7B%20type%3A%20'action1'%20%7D%20%2F%2F%20error%0Aconst%20action5%3A%20Action%3CVoid%3E%20%3D%20%7B%20type%3A%20'action1'%2C%20payload%3A%20undefined%20%7D%20%2F%2F%20error)

**Comment by gcanti** [comment]
@emirotin I'm not sure how the playground works (which version, which flags...), I found it unreliable: it often shows different results with respect to my local configuration. The code I posted seems to work fine in my setup (TypeScript v2.2.2)

**Comment by emirotin** [comment]
Huh, will try locally then, thanks again
